### PR TITLE
Asc 1277/master/mechanism for configurable test execution hierarchy

### PR DESCRIPTION
**Pull request by mwbutcher**

### CODE REVIEW

File: zigzag/data/configs/zigzag-config-example.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+    "test_cycle": "test_cycle_name",
```
Comment (zreichert) [inline zigzag/data/configs/zigzag-config-example.json]: Does this 'test_cycle' config overlap with the CLI option `--qtest-test-cycle` ?  If so which one is authoritative?

File: tests/unit/test_cli.py [hunk 2]
```diff
@@ -162,6 +162,123 @@ def test_cli_pprint_on_fail(missing_test_id_xml, mocker):
...
+
+    # Test
+    result = runner.invoke(cli.main, args=cli_arguments, env=env_vars)
+    assert 0 == result.exit_code
+    assert 'Queue Job ID: {}'.format(job_id) in result.output
+    assert 'Success!' in result.output
+
+
+def test_cli_malformed_config(invalid_config_file, single_passing_xml, mocker):
+    """Verify that the CLI will allow the user to set the '--zigzag_config_file' option."""
```
Comment (zreichert) [inline tests/unit/test_cli.py]: This description is not accurate.

File: tests/unit/test_cli.py [hunk 3]
```diff
@@ -162,6 +162,123 @@ def test_cli_pprint_on_fail(missing_test_id_xml, mocker):
...
+    mocker.patch('requests.post', return_value=mock_post_response)
+    mocker.patch('swagger_client.ObjectlinkApi.link_artifacts', return_value=[mock_link_response])
+
+    # Test
+    result = runner.invoke(cli.main, args=cli_arguments, env=env_vars)
+    assert 'config file is not valid JSON' in result.output
+
+
+def test_cli_missing_config(valid_config_file, single_passing_xml, mocker):
+    """Verify that the CLI will allow the user to set the '--zigzag_config_file' option."""
```
Comment (zreichert) [inline tests/unit/test_cli.py]: This description is not accurate

File: zigzag/data/configs/zigzag-config-example.json [hunk 4]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (zreichert) [inline zigzag/data/configs/zigzag-config-example.json]: I think we should have documentation on what all the possible config options are.

File: zigzag/data/configs/zigzag-config-example.json [hunk 5]
```diff
@@ -0,0 +1,5 @@
...
+    "test_cycle": "test_cycle_name",
+    "module_hierarchy": "{{ MODULE_HIERARCHY }}",
+    "path_to_test_exec_dir": "{{ PATH_TO_TEST_EXEC_DIR }}"
```
Comment (zreichert) [inline zigzag/data/configs/zigzag-config-example.json]: It been too long for me to remember if we have answered this already but .....
Is there a way to reference properties in the XML?  Things that may have been captured by an earlier tool?  (pytest-zigzag)

Comment (zreichert) [reply zigzag/data/configs/zigzag-config-example.json]: For example I saw this in the pytest-zigzag PR 
`"MODULE_HIERARCHY": {"type": ["array", "null"] },`
How would we use that value here in module_hierarchy?

Comment (mwbutcher) [reply zigzag/data/configs/zigzag-config-example.json]: I need to write this up in the docs, but yeah, you can reference any property you've specified in the xml.  So, to plumb it though, you can add an arbitrary_value to the config for pytest-zigzag, it will become a property in the resulting junit.xml, and you can interpolate it into the zigzag config by adding {{ arbitrary_value }}.

**Comment by mwbutcher** [comment]
added 'do not merge' to consider implications of recent pytest_zigzag work.

File: zigzag/data/configs/zigzag-config-example.json (hunk 5 above)
Comment (ryan-rs) [reply zigzag/data/configs/zigzag-config-example.json]: What does the `path_to_test_exec_dir` do exactly?

File: zigzag/data/molecule_junit.xsd [hunk 6]
```diff
@@ -45,7 +45,7 @@
     <!--Note: since sometimes either 'start_time' or 'end_time' might not be present because of setup/teardown
     failure we allow one less than the five possible test case properties.-->
     <xs:complexType name="testCasePropertiesType">
-        <xs:sequence minOccurs="4" maxOccurs="unbounded">
+        <xs:sequence minOccurs="0" maxOccurs="unbounded">
```
Comment (ryan-rs) [inline zigzag/data/molecule_junit.xsd]: Why was this changed to zero?

File: requirements.txt [hunk 7]
```diff
@@ -14,3 +14,4 @@ requests
 rpc-qtest-swagger-client
 tox
 twine
+jsonschema
```
Comment (ryan-rs) [inline requirements.txt]: This isn't used anywhere.

File: README.rst [hunk 8]
```diff
@@ -21,7 +21,28 @@ Quick Start Guide
...
+
+    Because the template is rendered using jinja2, arbitrary python
+    code can be used to inform the values. You could have a module hierarchy with
+    one of the nodes set to the datetime of the zigzag execution, for instance.
+
+    Presently there are three supported config properties:
+        test_cycle: A string name of the root node of the hierarchy for storing test results in qtest.
+        module_hierarchy: A list, of length => 0, of hierarchical nodes where test results will be stored in qtest.
+        path_to_test_exec_dir: A string representing an arbitrary path between the root of the project being tested and the directory
+            where tests will be executed. This is used
```
Comment (ryan-rs) [inline README.rst]: Looks like there should have been more text.

File: zigzag/cli.py [hunk 9]
```diff
@@ -23,13 +23,17 @@
               type=click.STRING,
               default=None,
               help='Specify a test cycle to use as a parent for results.')
+@click.option('--zigzag_config_file', '-c',
+              type=click.Path(exists=True),
+              default=None,
+              help='Path to a zigzag config file')
```
Comment (ryan-rs) [inline zigzag/cli.py]: What happens when the user doesn't supply the `--zigzag_config_file` option?

File: zigzag/module_hierarchy_facade.py [hunk 10]
```diff
@@ -32,7 +33,11 @@ def get_module_hierarchy(self, classname):
             list[str]: The strings to use for the module_hierarchy
         """
 
-        if self._mediator.ci_environment == 'asc':
+        if 'module_hierarchy' in self._mediator._config_dict.keys():
+            defined_hierarchy = ast.literal_eval(self._mediator._config_dict['module_hierarchy'])
```
Comment (ryan-rs) [inline zigzag/module_hierarchy_facade.py]: Eek! Why are we using a literal eval? We should just validate that the config file complies with a schema.

File: zigzag/data/configs/zigzag-config-example.json (hunk 5 above)
Comment (mwbutcher) [reply zigzag/data/configs/zigzag-config-example.json]: This is to handle the issue that @zreichert ran into with mk8s.  If a repo has tests in a nested dir, and they are executed from there, it's tough to put together a failure link to the problematic line in the repo without knowledge about the intervening path. The path_to_test_exec_dir property just stores that info if it's relevant.  If it's not defined, it should be ignored.

File: zigzag/data/molecule_junit.xsd (hunk 6 above)
Comment (mwbutcher) [reply zigzag/data/molecule_junit.xsd]: I was thinking that was too restrictive, but on second thought, i guess that we'll need: test_step, test_id, jira, and start_time at least...

File: requirements.txt (hunk 7 above)
Comment (mwbutcher) [reply requirements.txt]: oh, thanks!

File: zigzag/cli.py [hunk 11]
```diff
@@ -19,23 +19,15 @@
...
-              help='Specify a test cycle to use as a parent for results.')
-@click.option('--test-runner', '-tr',
-              type=click.Choice(['pytest-zigzag', 'tempest']),
-              default='pytest-zigzag',
-              help='Specify the tool that generated the xml to be processed')
+@click.argument('zigzag_config_file', type=click.Path(exists=True))
 @click.argument('junit_input_file', type=click.Path(exists=True))
-@click.argument('qtest_project_id', type=click.INT)
-def main(junit_input_file, qtest_project_id, qtest_test_cycle, pprint_on_fail, test_runner):
+def main(junit_input_file, zigzag_config_file, pprint_on_fail):
```
Comment (zreichert) [inline zigzag/cli.py]: nit: can you order these argument in the same order that they occur on the command line?  

Example:
`def main(zigzag_config_file, junit_input_file, pprint_on_fail):`

File: zigzag/cli.py [hunk 12]
```diff
@@ -19,23 +19,15 @@
...
     """Upload JUnitXML results to qTest manager.
 
     \b
     Required Arguments:
         JUNIT_INPUT_FILE        A valid JUnit XML results file.
-        QTEST_PROJECT_ID        The the target qTest Project ID for results
+        ZIGZAG_CONFIG_FILE      A valid json config file for zigzag.
```
Comment (zreichert) [inline zigzag/cli.py]: Can this also align with the order specified on the CLI

File: zigzag/module_hierarchy_facade.py [hunk 13]
```diff
@@ -32,31 +32,24 @@ def get_module_hierarchy(self, classname):
...
-            return self._mediator.testsuite_props['RPC_PRODUCT_RELEASE']
-        elif self._mediator.ci_environment == 'mk8s':
-            # organize all PR related testing by pr number ex: 'PR-123'
-            # organize all non PR related testing by branch name
-            return 'PR' if self.pr_match else self._mediator.testsuite_props['BRANCH_NAME']
-        elif self._mediator.test_runner == 'tempest':
-            return 'Tempest'
-        else:
-            return self._mediator.testsuite_props['RPC_PRODUCT_RELEASE']
+        return self._mediator._config_dict['test_cycle']
```
Comment (zreichert) [inline zigzag/module_hierarchy_facade.py]: YAY! less special case

File: zigzag/module_hierarchy_facade.py [hunk 14]
```diff
@@ -32,31 +32,24 @@ def get_module_hierarchy(self, classname):
             list[str]: The strings to use for the module_hierarchy
         """
 
-        if self._mediator.ci_environment == 'asc':
-            return self._asc(classname)
```
Comment (zreichert) [inline zigzag/module_hierarchy_facade.py]: I think we can remove these private methods.  `self._asc()` .....

File: zigzag/xml_parsing_facade.py [hunk 15]
```diff
@@ -65,7 +65,7 @@ def _determine_ci_environment(self):
         except AttributeError:
             self._mediator.ci_environment = 'asc'  # hard code this here
 
-    def _read(self):
+    def _read(self, file_path):
```
Comment (zreichert) [inline zigzag/xml_parsing_facade.py]: Can you update docstring to include file path or change it back so that it gets the file path from the mediator.

File: zigzag/zigzag.py [hunk 16]
```diff
@@ -37,12 +38,13 @@ def __init__(self,
         swagger_client.configuration.api_key['Authorization'] = qtest_api_token
         self._qtest_api_token = qtest_api_token
         self._junit_xml_file_path = junit_xml_file_path
-        self._qtest_project_id = qtest_project_id
-        self._qtest_test_cycle_name = qtest_test_cycle
         self._pprint_on_fail = pprint_on_fail
+        self._config_file = config_file
```
Comment (zreichert) [inline zigzag/zigzag.py]: I think you should create a property for config file to line up with other properties on this class.  Rather than accessing this internal attribute from outside this class.

File: zigzag/zigzag.py [hunk 17]
```diff
@@ -37,12 +38,13 @@ def __init__(self,
...
         self._test_logs = []
+        self._config_dict = {}
```
Comment (zreichert) [inline zigzag/zigzag.py]: create property for `config_dict`

File: zigzag/xml_parsing_facade.py (hunk 15 above)
Comment (zreichert) [reply zigzag/xml_parsing_facade.py]: I'll ticket this (lazy load cache pattern)

File: README.rst [hunk 18]
```diff
@@ -21,23 +21,33 @@ Quick Start Guide
...
-3. Here is an example of uploading a results JUnitXML file from Molecule::
+3. Configuring Zigzag::
+    The zigzag config file is defined by the json schema at:
+    zigzag/data/schema/zigzag-config.schema.json which is generated using a jinja2
+    template. (See zigzag/data/configs/zigzag-config-example.json for an example)
 
-    $ export QTEST_API_TOKEN="SECRET"
-    $ zigzag junit.xml 12345
+    Any test suite property from the incoming junit.xml can be interpolated, by
+    name, into the template. (See the example.)
```
Comment (ryan-rs) [inline README.rst]: The formatting is a little off in this section. Also, you should make `zigzag/data/configs/zigzag-config-example.json` a hyperlink.

File: tests/unit/conftest.py [hunk 19]
```diff
@@ -131,6 +179,65 @@ def single_passing_xml(tmpdir_factory,
...
+           """
+
+    with open(filename, 'w') as f:
+        f.write(config_json)
+
+    return filename
+
+
+@pytest.fixture(scope='session')
+def valid_config_file(tmpdir_factory,
```
Comment (ryan-rs) [inline tests/unit/conftest.py]: This is a 'pytest-zigzag-config' file. The name and description of the fixture is misleading.
